### ISSUE TITLE
postgres: `create procedure` with `out` parameter detected as unparseable section

### ISSUE THREAD

**Issue by BrunoJuchli** [sqlfluff/sqlfluff#4986]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When linting / fixing the following an SQL statement (postgres dialect) with a `create procedure` I get the error: 

>  Found unparsable section: 'CREATE PROCEDURE LmMultilevelFetchNextSh...'

I noticed that removing the `out` parameter makes the linting work.

### Expected Behaviour

The procedure is expected to be detected and its contents ignored. The rest of the code should be linted.
No error should be reported.

### Observed Behaviour

Fails with:
>L:   8 | P:   1 |  PRS | Line 8, Position 1: Found unparsable section: 'create
                       | procedure LmMultilevelFetchNextSh...'

### How to reproduce

`sqlfluff lint` the following file:

    create table LmMultilevelShardStates (
        Shard smallint primary key,
        LastReserved timestamp default '2000-01-01 00:00:00'
    );

    insert into LmMultilevelShardStates (Shard) select * from generate_series(1, 2048);

    create procedure LmMultilevelFetchNextShards
    (
        batchSize smallint,
        timeout interval,
        shards out smallint[]
    )
    as $$
    declare
        now timestamp := NOW();
        timedout timestamp := now - timeout;
    begin  
        with reserved_shards as (
            update LmMultilevelShardStates
                set LastReserved = now
                where Shard in (
                    select Shard
                    from LmMultilevelShardStates
                    where LmMultilevelShardStates.LastReserved <= timedout
                    order by LmMultilevelShardStates.LastReserved
                    limit batchSize
                    for update)
                returning Shard)
        select array_agg(shard)
        from reserved_shards
        into shards;    
    end;
    $$ language plpgsql;

### Dialect

postgres

### Version

2.1.2

### Configuration

.sqlfluff:

    [sqlfluff]
    dialect = postgres
    templater = raw
    exclude_rules = layout.end_of_file, ambiguous.column_count, structure.column_order
    max_line_length = 120
    processes = 0

    [sqlfluff:indentation]
    allow_implicit_indents = true

    [sqlfluff:rules:aliasing.table]
    aliasing = explicit
    [sqlfluff:rules:aliasing.column]
    aliasing = explicit
    [sqlfluff:rules:aliasing.length]
    min_alias_length = 3

    [sqlfluff:rules:capitalisation.keywords]
    capitalisation_policy = lower
    [sqlfluff:rules:capitalisation.identifiers]
    capitalisation_policy = lower
    [sqlfluff:rules:capitalisation.functions]
    extended_capitalisation_policy = lower
    [sqlfluff:rules:capitalisation.literals]
    capitalisation_policy = lower
    [sqlfluff:rules:capitalisation.types]
    extended_capitalisation_policy = lower

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by BrunoJuchli** [comment]
## Workaround

I noticed that changing the `out` keyword to the front of the parameter definition fixed the issue.

### 😞 Bad

     shards out smallint []

### 😃 Good

     out shards smallint []

**Comment by joaostorrer** [comment]
From the [Postgres documentation](https://www.postgresql.org/docs/16/sql-createprocedure.html), the `argmode` (The mode of an argument: IN, OUT, INOUT, or VARIADIC) should be the first argument, but somehow Postgres is accepting it as second argument (I tested against a Postgres 15 database).

**Comment by alpha-beta-soup** [comment]
I get an `unparseable section 'VARIADIC ARRAY_REMOVE(ARRAY[` for 

```sql
range_merge(datemultirange(
    VARIADIC ARRAY_REMOVE(ARRAY[
        a.source_date,
        b.source_date,
        c.source_date
    ], NULL)
))::daterange, -- source_date
```

`datemultirange` does not accept an array, so `VARIADIC` declares the argmode so I can pass an array. The alternative is much uglier:

```sql
range_merge(
    CASE 
        WHEN a.source_date IS NOT NULL AND b.source_date IS NOT NULL AND c.source_date IS NOT NULL
        THEN datemultirange(a.source_date, b.source_date, c.source_date)
        WHEN a.source_date IS NOT NULL AND b.source_date IS NOT NULL
        THEN datemultirange(a.source_date, b.source_date)
        WHEN a.source_date IS NOT NULL AND c.source_date IS NOT NULL
        THEN datemultirange(a.source_date, c.source_date)
        WHEN b.source_date IS NOT NULL AND c.source_date IS NOT NULL
        THEN datemultirange(b.source_date, c.source_date)
        WHEN a.source_date IS NOT NULL THEN datemultirange(a.source_date)
        WHEN b.source_date IS NOT NULL THEN datemultirange(b.source_date)
        WHEN c.source_date IS NOT NULL THEN datemultirange(c.source_date)
        ELSE NULL
    END
)::daterange AS source_date
```

I'm not sure whether this is better as a separate issue, it's just the only hit when I searhc existing issues for `VARIADIC`.

**Comment by joaostorrer** [comment]
@alpha-beta-soup definitely the case you're describing is a separate issue. Can you open a new issue with this example?

**Comment by franloza** [comment]
I opened a PR to fix a related issue (https://github.com/sqlfluff/sqlfluff/issues/6784), but I could not reproduce this issue with version `3.4.1`

```bash
$ sqlfluff --version
sqlfluff, version 3.4.1

$ sqlfluff lint test_procedure.sql --dialect postgres
== [test_procedure.sql] FAIL                                       
L:   1 | P:  14 | CP02 | Unquoted identifiers must be consistently upper case.
                       | [capitalisation.identifiers]
L:   2 | P:   5 | CP02 | Unquoted identifiers must be consistently upper case.
                       | [capitalisation.identifiers]
L:   3 | P:   5 | CP02 | Unquoted identifiers must be consistently upper case.
                       | [capitalisation.identifiers]
L:   6 | P:  13 | CP02 | Unquoted identifiers must be consistently upper case.
                       | [capitalisation.identifiers]
L:   6 | P:  38 | CP02 | Unquoted identifiers must be consistently upper case.
                       | [capitalisation.identifiers]
L:   8 | P:  18 | CP03 | Function names must be lower case.
                       | [capitalisation.functions]
L:   8 | P:  45 | LT01 | Unexpected line break. [layout.spacing]
L:  12 | P:  24 | LT01 | Expected single whitespace between 'smallint' keyword
                       | and start square bracket '['. [layout.spacing]
L:  34 | P:  13 | CP02 | Unquoted identifiers must be consistently upper case.
                       | [capitalisation.identifiers]
All Finished 📜 🎉!
```

@BrunoJuchli Could you confirm this is an ongoing issue?

**Comment by greg-finley** [comment]
This parses fine now! https://online.sqlfluff.com/fluffed?sql=H4sIAN6iBWkC_41TwU7DMAy9V-o_-FBpHWKoIA6IwZXTGBI7cEBo8hrTRaRpcVJG-XrSpNsoYxJVLvGzn_2e05wJLYHFlSKYlfeNslLRB6nFGlksrAMNpHEE7vMhMCUqJbWFmmWJ3MIbtachYYbGPpIh_iABVpZkLJY1CHpFxwujiyzLJtm5O5Bl1_6M4mg8jaM4ktrVWXDE1fE5_GUMhhTlFk7glasSCtLEDl86AkkmPT-Fi-zyKtDmQV_NVU6i4aHGO7L5ek6f1vOaOOqFrtDFF_KLdmJ7gZ2mqvFTOpGo-rDx5dAh24LnFycsjtBAksSRoFwhU8jW1eaHOde3MH94Srtptx1ERzTI6Eom2-4uc0WF1AChZCPtGri3fdnPgrutdV9Ti86FI77u87wYssNF-vbDnM2anJXhPbhB0iEaWPyKfMoh6vf2r2n23Y6knw1mvbndOXjIU7EghlX7L6rDciVLafdv4w9VFfdOj4cgk21YS10EP3q0twiZsV1iUaTmB-gd-rXUgPg_JASm3T2OSAv3JpIEFOqiwcI9d1UX5l1NvwEiaDaF3gMAAA%3D%3D&dialect=PostgreSQL
